### PR TITLE
avoid creating temporary in genObjConstr if possible

### PR DESCRIPTION
**Pull request by jcosborn**
I've found this to give over a 2x improvement in some code that constructs nontrivial objects.

### CODE REVIEW

File: compiler/ccgexprs.nim [hunk 1]
```diff
@@ -1204,7 +1204,11 @@ proc genObjConstr(p: BProc, e: PNode, d: var TLoc) =
     if handleConstExpr(p, e, d): return
   var tmp: TLoc
   var t = e.typ.skipTypes(abstractInst)
-  getTemp(p, t, tmp)
+  let noTemp = d.k in {locTemp,locLocalVar,locGlobalVar,locParam,locField}
```
Comment (Araq) [inline compiler/ccgexprs.nim]: I doubt this is correct, what if I do ``x = Obj(field: x)``? (That's valid for ``ref object`` at least.)

**Comment by jcosborn** [comment]
Ah, yes.  I reverted the changes and added a test case for a self-referencing assignment.  Is there a good way to detect this?

**Comment by Araq** [comment]
> Is there a good way to detect this?

In ccgexprs.nim there is 

```nim

proc lhsDoesAlias(a, b: PNode): bool =
  for y in b:
    if isPartOf(a, y) != arNo: return true
``` 

Which uses alias analysis to detect it.

**Comment by Araq** [comment]
Appveyor doesn't like this:

Error: invalid command line option: '--nilchecks'

**Comment by jcosborn** [comment]
Yes, linux works, but windows fails.  I don't have windows to check it.  Are the csources versions different?
Avoiding nilchecks makes a big difference and it is only used in resetLoc.  Can I just disable it there for now?

**Comment by Araq** [comment]
> Yes, linux works, but windows fails. I don't have windows to check it. Are the csources versions different?

I don't think so, I think appveyor's setup is simply different.

> Avoiding nilchecks makes a big difference and it is only used in resetLoc. Can I just disable it there for now?

This configuration change is not part of the PR anyway, leave it out for now? Or you add a ``nim.cfg`` just for this test.

**Comment by jcosborn** [comment]
Ok, Appveyor bootstraps with -d:release, Travis doesn't.
I've removed nilchecks from the config, so hopefully everything builds now, though the performance may actually be slower unless one compiles with --nilchecks:off.
I'll make a separate PR which makes that default to off.

**Pull request merged**